File: examples/swift/records.swift
import Foundation

@main
struct RecordsExample {
    static func main() {
        do {
            try run()
        } catch {
            fatalError("Swift records example failed: \(error)")
        }
    }

    static func run() throws {
        let reader = try reader_from_bytes(decodeSample())
        let stream = try reader_records(reader)

        var counts: [String: Int] = [
            "Header": 0,
            "Footer": 0,
            "Schema": 0,
            "Channel": 0,
            "Message": 0,
            "Chunk": 0,
            "Other": 0,
        ]

        while true {
            let item = try record_stream_next(stream)
            if !item.has_value {
                break
            }
            let record = item.value
            let kind = record.kind.toString()
            let key = counts.keys.contains(kind) ? kind : "Other"
            counts[key, default: 0] += 1
        }

        print("Record counts:")
        for key in ["Header", "Footer", "Schema", "Channel", "Message", "Chunk", "Other"] {
            print("  \(key): \(counts[key, default: 0])")
        }
    }
}

File: examples/rust/records.rs
//! Low-level record iteration.
//!
//! Run with: cargo run -p mcapable --example records

mod common_bytes;
mod common_file;

use mcapable::{Record, reader};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = common_file::write_sample_file()?;
    let mut reader = reader::Builder::new().build(file.reopen()?)?;

    // Count record types
    let mut header_count = 0;
    let mut footer_count = 0;
    let mut schema_count = 0;
    let mut channel_count = 0;
    let mut message_count = 0;
    let mut chunk_count = 0;
    let mut other_count = 0;

    for record in reader.records() {
        match record? {
            Record::Header(_) => header_count += 1,
            Record::Footer(_) => footer_count += 1,
            Record::Schema(_) => schema_count += 1,
            Record::Channel(_) => channel_count += 1,
            Record::Message(_) => message_count += 1,
            Record::Chunk(_) => chunk_count += 1,
            _ => other_count += 1,
        }
    }

    println!("Record counts:");
    println!("  Header:   {}", header_count);
    println!("  Footer:   {}", footer_count);
    println!("  Schema:   {}", schema_count);
    println!("  Channel:  {}", channel_count);
    println!("  Message:  {}", message_count);
    println!("  Chunk:    {}", chunk_count);
    println!("  Other:    {}", other_count);

    Ok(())
}
